File: ports/python/confine_core/executor.py
"""Executor conformance (spec PROTOCOL.md sec 11) -- the normative
invariant this entire project is built around.

A conforming executor MUST independently invoke the complete policy
decision procedure over the exact current state, canonical action, actor
identity, capability set, and policy object, immediately before
committing any effect. A certificate, its obligations, or a previously
computed policy result MUST NOT substitute for this invocation.
"""

from .certificate import verify as certificate_verify, _obligation_to_dict
from .policy import verify as policy_verify
from .state import nonce_consumed
from .canonical import encode


class ExecuteError(Exception):
    def __init__(self, code, policy_code=None):
        self.code = code
        self.policy_code = policy_code
        super().__init__(code)


def execute_exact(state, action, actor_id, capabilities, policy, certificate, broker_secret):
    cv = certificate_verify(certificate, state, action, actor_id, capabilities, policy, broker_secret)
    if cv is not None:
        raise ExecuteError(cv.code)

    decision = policy_verify(state, action, actor_id, capabilities, policy)
    if decision["t"] == "deny":
        raise ExecuteError("EXECUTOR_POLICY_DENIED", policy_code=decision["code"])

    real_obligations = decision["obligations"]
    cert_obligations_encoded = encode([_obligation_to_dict(o) for o in certificate.obligations])
    real_obligations_encoded = encode([_obligation_to_dict(o) for o in real_obligations])
    if cert_obligations_encoded != real_obligations_encoded:
        raise ExecuteError("CERTIFICATE_OBLIGATION_MISMATCH")

    if nonce_consumed(state, action["nonce"]):
        raise ExecuteError("NONCE_REPLAY_AT_EXECUTOR")

    return {"t": "allow", "obligations": real_obligations}

File: ports/js/src/executor.js
// Executor conformance (spec PROTOCOL.md sec 11) -- the normative
// invariant this entire project is built around.
//
// A conforming executor MUST independently invoke the complete policy
// decision procedure over the exact current state, canonical action,
// actor identity, capability set, and policy object, immediately before
// committing any effect. A certificate, its obligations, or a previously
// computed policy result MUST NOT substitute for this invocation.

import { verify as certificateVerify, obligationToValue } from "./certificate.js";
import { verify as policyVerify } from "./policy.js";
import { nonceConsumed } from "./state.js";
import { encode } from "./canonical.js";

export class ExecuteError extends Error {
  constructor(code, policyCode = null) {
    super(code);
    this.code = code;
    this.policyCode = policyCode;
  }
}

export function executeExact(state, action, actorId, capabilities, policy, certificate, brokerSecret) {
  const cv = certificateVerify(certificate, state, action, actorId, capabilities, policy, brokerSecret);
  if (cv !== null) {
    throw new ExecuteError(cv);
  }

  const decision = policyVerify(state, action, actorId, capabilities, policy);
  if (decision.t === "deny") {
    throw new ExecuteError("EXECUTOR_POLICY_DENIED", decision.code);
  }

  const certObligationsEncoded = encode(certificate.obligations.map(obligationToValue));
  const realObligationsEncoded = encode(decision.obligations.map(obligationToValue));
  if (certObligationsEncoded !== realObligationsEncoded) {
    throw new ExecuteError("CERTIFICATE_OBLIGATION_MISMATCH");
  }

  if (nonceConsumed(state, action.nonce)) {
    throw new ExecuteError("NONCE_REPLAY_AT_EXECUTOR");
  }

  return { t: "allow", obligations: decision.obligations };
}
